mdpow/analysis/ensemble.py

```diff
@@ -0,0 +1,499 @@
+        for i in range(len(topologies)):
+            file = topologies[i]
+            if file.endswith('.tpr'):
+                topologies.pop(i)
+                top = [file] + topologies
+                break
+        return top
+
+    def _load_universe_from_dir(self, solvent):
+        """Loads system in directory into an MDAnalysis Universe
+
+        logs warning if more than one topology is in directory. If
+        more than one trajectory attempts to load both of them
+        in a universe if that fail will try to load each individually"""
+        cur_dir = os.listdir(os.curdir)
+        self.trj = []
+        self.top = []
+
+        if not self.top_dict is None:
+            # if top is specified in kwargs, saved to list
+            self.top = [self.top_dict[solvent]]
+
+        for file in cur_dir:
+            if file.endswith('.xtc'):
+                # Saving trajectory directories
+                self.trj.append(file)
+            elif (file.endswith('gro') or file.endswith('.tpr') or file.endswith('gro.b2z') or file.endswith('gro.gz'))\
+                    and self.top_dict is None:
+                # Saving topology directories
+                self.top.append(file)
+
+        if len(self.top) == 0 or len(self.trj) == 0:
+            logger.warning('No MD files detected in %s', os.curdir)
+            raise NoDataWarning
+
+        self.trj = self._sort_trajectories(self.trj)
+        if len(self.top) > 1:
+            self.top = self._sort_topologies(self.top)
+
+    def keys(self):
+        """Returns list of system keys"""
+        return self._keys
+
+    def _build_ensemble(self):
+        """Goes through given mdpow directory and attempts to build universe in the lambda directories.
+        Run if dirname is passed into __init__. First enters FEP directory, then goes through solvent
+        and interaction directories to search lambda directories for system files."""
+        fep_dir = os.path.join(self._ensemble_dir, 'FEP')
+        for solvent in self._solvents:  # Ugly set of loops, may have to find way to clean up
+            for dirs in self._interactions:  # Attribute folder names
+                int_dir = os.path.join(fep_dir, solvent, dirs)
+                if os.path.exists(int_dir):
+                    with in_dir(int_dir, create=False):  # Entering attribute folders
+                        logger.info("Searching %s directory for systems", os.curdir)
+                        files = os.listdir(os.curdir)
+                        for file in files:  # Traversing lambda directories
+                            if os.path.isdir(file):
+                                with in_dir(file, create=False):
+                                    try:
+                                        self._load_universe_from_dir(solvent)
+                                        self.add_system((solvent, dirs, file), topology=self.top[0],
+                                                        trajectory=self.trj)
```

Comment:
Change to something along the lines of
```python
u = self._load_universe_from_dir(solvent=solvent)
if u is None:
   logger.warn(...)
   warnings.warn(..., category=MissingDataWarning)
key = (solvent, dirs, file)
self.add_system(key, u)
```

If there are exceptions to be caught then better use individual try/except blocks.

Reply:
Done, now all the errors and exceptions are done in `_load_unviverse_from_dir()`